modules/phase_field/include/kernels/ACGrGrMulti.h

```diff
@@ -10,37 +10,63 @@
 #pragma once
 
 #include "ACGrGrBase.h"
-
-// Forward Declarations
+#include "ADGrainGrowthBase.h"
 
 /**
  * This kernel calculates the residual for grain growth for a multi-phase,
  * poly-crystal system. A list of material properties needs to be supplied for the gammas
  * (prefactors of the cross-terms between order parameters).
  */
-class ACGrGrMulti : public ACGrGrBase
+
+template <bool is_ad>
+using ACGrGrMultiBase = typename std::conditional<is_ad, ADGrainGrowthBase, ACGrGrBase>::type;
+
+template <bool is_ad>
+class ACGrGrMultiTempl : public ACGrGrMultiBase<is_ad>
 {
 public:
   static InputParameters validParams();
 
-  ACGrGrMulti(const InputParameters & parameters);
+  ACGrGrMultiTempl(const InputParameters & parameters);
 
 protected:
-  virtual Real computeDFDOP(PFFunctionType type);
-  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
-
   /// Names of gammas for each order parameter
   std::vector<MaterialPropertyName> _gamma_names;
   unsigned int _num_j;
 
   /// Values of gammas for each order parameter
-  std::vector<const MaterialProperty<Real> *> _prop_gammas;
+  std::vector<const GenericMaterialProperty<Real, is_ad> *> _prop_gammas;
+
+  using ACGrGrMultiBase<is_ad>::_op_num;
+  using ACGrGrMultiBase<is_ad>::_qp;
+  using ACGrGrMultiBase<is_ad>::_vals;
+  using ACGrGrMultiBase<is_ad>::_u;
+  using ACGrGrMultiBase<is_ad>::_mu;
+  using ACGrGrMultiBase<is_ad>::_test;
+
+  GenericReal<is_ad> computedF0du();
+};
+
+class ACGrGrMulti : public ACGrGrMultiTempl<false>
+{
+public:
+  ACGrGrMulti(const InputParameters & parameters);
+
+protected:
+  virtual Real computeDFDOP(PFFunctionType type);
```

Comment:
Why does this take an argument and the AD version doesn't?

Reply:
Because the non-AD version calculates both residual and jacobian, that's the type inhput here.